lib/aws-dev-utils/next_token_wrapper.rb

```diff
@@ -9,23 +9,30 @@ def initialize client, max=100
       @max = max
     end
 
-    def method_missing m, *args, &block
-      do_call(m,*args,&block)
+    def method_missing method, *args, &block
+      do_call(method, *args, &block)
     end
 
     private
-
-    def do_call m, *args, &block
-      @client.send(m , *args, &block).to_h.tap do |r|
+    def do_call method, *args, &block
+      @client.send(method, *args, &block).to_h.tap do |response|
         i = 1
-        key, key2 = extract_keys r
-        while(key && r[key] && i < @max) do
+        resp_keys, req_keys = extract_keys response
+
+        resp_keys = Array(resp_keys)
+        req_keys = Array(req_keys)
+
+        next response if resp_keys.empty?
+
+        props = args.first || {}
+        while resp_keys.all?{ |resp_key| response[resp_key] } && i < @max do
           i += 1
-          res = @client.send(m, (args[0]||{}).merge(key2 => r[key])).to_h
-          res.each { |k,v| r[k] = v.is_a?(Array) ? r[k].concat(v) : v }
-          r.delete_if {|k,v| !res.keys.include?(k) }
+          pagination_token_props =  Hash[req_keys.zip(resp_keys.map { |resp_key| response[resp_key] } ) ]
+          new_response = @client.send(method, props.merge(pagination_token_props)).to_h
+          new_response.each { |k,v| response[k] = v.is_a?(Array) ? response[k].concat(v) : v }
+          response.delete_if { |k,v| !new_response.keys.include?(k) }
```

Comment:
it seems that you get the new_response and you merge it to the old response and delete the keys not in the new response, why not use the new response instead of the old and merge the values from the old response to the new response?